### PR TITLE
[SPARK-7068][SQL] Remove PrimitiveType

### PR DESCRIPTION
**Pull request by rxin**

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/types/dataTypes.scala [hunk 1]
```diff
@@ -41,6 +41,21 @@ import org.apache.spark.util.Utils
 object DataType {
   def fromJson(json: String): DataType = parseDataType(parse(json))
 
+  private val nonDecimalNameToType = {
```
Comment (rxin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/types/dataTypes.scala]: this is just copied from down below (originally in PrimitiveType)

File: sql/catalyst/src/main/scala/org/apache/spark/sql/types/dataTypes.scala [hunk 2]
```diff
@@ -190,13 +205,11 @@ object DataType {
...
         equalsIgnoreNullability(leftValueType, rightValueType)
       case (StructType(leftFields), StructType(rightFields)) =>
-        leftFields.size == rightFields.size &&
-        leftFields.zip(rightFields)
-          .forall{
-            case (left, right) =>
-              left.name == right.name && equalsIgnoreNullability(left.dataType, right.dataType)
-          }
-      case (left, right) => left == right
+        leftFields.length == rightFields.length &&
```
Comment (rxin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/types/dataTypes.scala]: left / right were shadowing outer scope variables so i renamed them l, r

File: sql/core/src/main/scala/org/apache/spark/sql/parquet/ParquetTypes.scala [hunk 3]
```diff
@@ -48,8 +48,10 @@ private[parquet] case class ParquetTypeInfo(
   length: Option[Int] = None)
 
 private[parquet] object ParquetTypesConverter extends Logging {
-  def isPrimitiveType(ctype: DataType): Boolean =
-    classOf[PrimitiveType] isAssignableFrom ctype.getClass
+  def isPrimitiveType(ctype: DataType): Boolean = ctype match {
+    case _: NumericType | BooleanType | StringType | BinaryType => true
```
Comment (liancheng) [inline sql/core/src/main/scala/org/apache/spark/sql/parquet/ParquetTypes.scala]: We need to put all types other than `MapType`, `ArrayType`, and `StructType` here. Parquet uses this function to create record converters [here] [1].

[1]: https://github.com/apache/spark/blob/03e85b4a11899f37424cd6e1f8d71f1d704c90bb/sql/core/src/main/scala/org/apache/spark/sql/parquet/ParquetConverter.scala#L161-L165

Comment (liancheng) [inline sql/core/src/main/scala/org/apache/spark/sql/parquet/ParquetTypes.scala]: My bad. `CatalystGroupConverter` actually can handle all types properly. Thought it only handles complex types at first.

**Comment by liancheng** [comment]
LGTM pending Jenkins.

**Pull request closed without merging**